### PR TITLE
I3336 webhook namespace scoping

### PR DESCRIPTION
**Pull request by mlholland**
# Description

The change validation webhook respected the `WatchNamespaces` value for filtering, but not the readGatewaysFromAllNamespaces value. This change adds the latter into consideration, with the following behavior for reading gateway changes based on the two values (note that when WatchNamespaces is empty, this means that it watches over all namespaces):
 - WatchNamespaces empty, readGatewaysFromAllNamespaces false: Only read gateway changes from the webhook's namespace
 - WatchNamespaces populated, readGatewaysFromAllNamespaces false: Only read gateway changes from the webhook's namespace
 - WatchNamespaces empty, readGatewaysFromAllNamespaces true: Read gateway changes from all namespaces
 - WatchNamespaces populated, readGatewaysFromAllNamespaces true: Only gateway changes from the WatchNamespaces list.

These changes make it necessary for the validation webhook to know its own namespace name, so a new parameter was added to account for this, which filtered all the way back to the setup file.

Questions for reviewers:
 - Are the changes in setup_syncer.go worth testing in a test file? If so, which file?
 - The changes in setup_syncer.go incidentally add a new restriction on valid arguments, should we document them somewhere, or is the error message produced sufficient?

# Context

https://github.com/solo-io/gloo/issues/3336

# Checklist:

- [X] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [X] If I updated APIs (our protos) or helm values, I ran `make install-go-tools generated-code` to ensure there will be no code diff
- [X] I followed guidelines laid out in the Gloo [contribution guide](https://docs.solo.io/gloo/latest/contributing/)
- [X] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/3336

### CODE REVIEW

File: projects/gateway/pkg/syncer/setup_syncer.go [hunk 1]
```diff
@@ -269,6 +269,23 @@ func RunGateway(opts translator.Opts) error {
...
+		if !opts.ReadGatewaysFromAllNamespaces && !utils.AllNamespaces(opts.WatchNamespaces) {
+			foundSelf := false
+			for _, namespace := range opts.WatchNamespaces {
+				if controllerNamespace == namespace {
+					foundSelf = true
+					break
+				}
+			}
+			if !foundSelf {
+				return errors.Errorf("READ_GATEWAYS_FROM_ALL_NAMESPACES was set to false, but non-empty WATCH_NAMESPACES"+
```
Comment (kdorosh) [inline projects/gateway/pkg/syncer/setup_syncer.go:284]: nit, `READ_GATEWAYS_FROM_ALL_NAMESPACES` and `WATCH_NAMESPACES` are not env vars, but rather live on the gloo settings. the error message can be improved to reflect this (for an example, see:

```golang
return errors.Errorf("VALIDATION_MUST_START was set to true, but no validation configuration was provided in the settings. "+
				"Ensure the v1.Settings %v contains the spec.gateway.validation config", settings.GetMetadata().Ref())
```
)

File: projects/gateway/pkg/syncer/setup_syncer.go [hunk 2]
```diff
@@ -269,6 +269,23 @@ func RunGateway(opts translator.Opts) error {
...
+					" list did not contain this gloo instance's own namespace: %s", controllerNamespace)
```
Comment (kdorosh) [inline projects/gateway/pkg/syncer/setup_syncer.go:285]: nit, we should log what `watchNamespaces` is set to

File: projects/gateway/pkg/syncer/setup_syncer.go [hunk 3]
```diff
@@ -269,6 +269,23 @@ func RunGateway(opts translator.Opts) error {
 
 	validationServerErr := make(chan error, 1)
 	if opts.Validation != nil {
+		// make sure non-empty WatchNamespaces contains the gloo instance's own namespace if
+		// ReadGatewaysFromAllNamespaces is false
+		controllerNamespace := os.Getenv("POD_NAMESPACE")
```
Comment (kdorosh) [inline projects/gateway/pkg/syncer/setup_syncer.go:274]: if we aren't running in k8s via helm (e.g. running gloo on files, consul, etc), then this env var will most often be unset and thus return empty. we may want an error thrown here if `POD_NAMESPACE` is unset

File: projects/gateway/pkg/syncer/setup_syncer.go [hunk 4]
```diff
@@ -269,6 +272,26 @@ func RunGateway(opts translator.Opts) error {
 
 	validationServerErr := make(chan error, 1)
 	if opts.Validation != nil {
+		// make sure non-empty WatchNamespaces contains the gloo instance's own namespace if
+		// ReadGatewaysFromAllNamespaces is false
+		controllerNamespace := os.Getenv(setuputils.PodNamespace)
```
Comment (kdorosh) [inline projects/gateway/pkg/syncer/setup_syncer.go:277]: i know it was originally my suggestion, but i think we should be able to do this instead (and thus remove the check following)
```suggestion
settings.Metadata.Namespace
```
as this should always equal `POD_NAMESPACE` (or the equivalent if passed by flag in CLI) -- even in the case of running gloo outside of k8s

Comment (mlholland) [reply projects/gateway/pkg/syncer/setup_syncer.go:277]: this is going to require some refactoring - the settings variable isn't passed into the useGateway function. This function  has one other usage, and I don't immediately see how to get the namespace in the other case (gateway.go)

**Comment by mlholland** [comment]
/kick this seems like a flake, will verify on both ends while i figure out how to run gateway test locally

File: projects/gateway/pkg/services/k8sadmisssion/validating_admission_webhook.go [hunk 5]
```diff
@@ -90,10 +90,20 @@ type WebhookConfig struct {
 	port                          int
 	serverCertPath, serverKeyPath string
 	alwaysAccept                  bool // accept all resources
+	readGatewaysFromAllNamespaces bool
+	webhookNamespace              string
 }
 
-func NewWebhookConfig(ctx context.Context, validator validation.Validator, watchNamespaces []string, port int, serverCertPath string, serverKeyPath string, alwaysAccept bool) WebhookConfig {
-	return WebhookConfig{ctx: ctx, validator: validator, watchNamespaces: watchNamespaces, port: port, serverCertPath: serverCertPath, serverKeyPath: serverKeyPath, alwaysAccept: alwaysAccept}
+func NewWebhookConfig(ctx context.Context, validator validation.Validator, watchNamespaces []string, port int, serverCertPath string, serverKeyPath string, alwaysAccept bool, readGatewaysFromAllNamespaces bool, webhookNamespace string) WebhookConfig {
```
Comment (kdorosh) [inline projects/gateway/pkg/services/k8sadmisssion/validating_admission_webhook.go:97]: fyi (not asking you to make this change and send through CI again) when declaring variables of the same type, you don't need to provide the type every time.

i.e.
```golang
func NewWebhookConfig(ctx context.Context, ..., alwaysAccept bool, readGatewaysFromAllNamespaces bool, webhookNamespace string) WebhookConfig {
```
can be

```golang
func NewWebhookConfig(ctx context.Context, ..., alwaysAccept, readGatewaysFromAllNamespaces bool, webhookNamespace string) WebhookConfig {
```

File: projects/gateway/pkg/syncer/setup_syncer.go [hunk 6]
```diff
@@ -146,10 +147,11 @@ func Setup(ctx context.Context, kubeCache kube.SharedCache, inMemoryCache memory
 		Validation:                    validation,
 	}
 
-	return RunGateway(opts)
+	return RunGateway(opts, settings.Metadata.Namespace)
 }
 
-func RunGateway(opts translator.Opts) error {
+// the need for the namespace is limited to this function, whereas the opts struct's use is more widespread.
+func RunGateway(opts translator.Opts, glooNamespace string) error {
```
Comment (kdorosh) [inline projects/gateway/pkg/syncer/setup_syncer.go:154]: nit, i think it's more natural to have `glooNamespace` in the `translator.Opts` struct

File: projects/gateway/pkg/services/k8sadmisssion/validating_admission_webhook_test.go [hunk 7]
```diff
@@ -146,6 +149,48 @@ var _ = Describe("ValidatingAdmissionWebhook", func() {
...
+
+			review, err := parseReviewResponse(res)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(review.Response).NotTo(BeNil())
+
+			Expect(review.Response.Allowed).To(BeTrue())
+			Expect(review.Response.Result).To(BeNil())
+		})
+
+		It("Does not process other-namespace gateway resources if readGatewaysFromAllNamespaces is false, even if they're from whitelisted namespaces", func() {
```
Comment (kdorosh) [inline projects/gateway/pkg/services/k8sadmisssion/validating_admission_webhook_test.go:173]: super nit, typically we've been doing lower case after `It(`

```suggestion
		It("does not process other-namespace gateway resources if readGatewaysFromAllNamespaces is false, even if they're from whitelisted namespaces", func() {
```

**Pull request merged**